### ISSUE TITLE
Unit mismatch for three-body reactions

### ISSUE THREAD

**Issue by jAnirudh** [AMReX-Combustion/PelePhysics#323]
Hi,

I am not sure but I may have found a bug in the reaction rate coefficients units of three body reactions.
 
I was looking at the Li-Dryer Hydrogen combustion mechanism which has 4 three-body reactions and found that CEPTR adds an additional _meter ** 3 / kilomole_ factor to each three-body reaction when compared to Cantera.

I've made some basic [changes](https://github.com/AMReX-Combustion/PelePhysics/commit/306ab2ba4d760687cfce9c3243ad4e66aeb61dc7) to `production.py` and wrote a script `Testing/Exec/3bodyTest/threebody_Aunits.py` to get this output: [units_dump.txt](https://github.com/AMReX-Combustion/PelePhysics/files/9692099/units_dump.txt). Essentially this difference in units increases the magnitudes of the forward reaction rates by a factor of 1000.

I tried to see the effect of this by looking at the pressure scaling results reported [here in (Fig8d) ](https://arc.aiaa.org/doi/10.2514/6.2021-3448) but this additional 1000 multiplying factor does not seem to affect the results as shown below.

![1bar](https://user-images.githubusercontent.com/7749537/193446812-ffb0835d-4ae3-490d-976c-0461802cea75.png)
![5bar](https://user-images.githubusercontent.com/7749537/193446844-185b7b42-65ba-4a47-83a1-7411c1a14209.png)
![10bar](https://user-images.githubusercontent.com/7749537/193446849-a5099fea-2f2b-47e0-91bf-46c553c7fab8.png)

The results I've shown here can be reproduced from `Testing/Exec/3bodyTest/` in [this fork](https://github.com/jAnirudh/PelePhysics/tree/3BodyRxnUnits). For `unit_dump.txt` run `python threebody_Aunits.py` and for the remaining plots run `sh exec.sh`. 

I also tried computing the ignition delay for this problem but it turns out that the temperature profile does not show any deviation from the expected profiles.

Again, I am not sure if this a bug, but thought it may be worth reporting.

Cheers,
Anirudh

**Comment by malihass** [comment]
Thanks for reporting this,  I finally got the chance to dive into this.

So I looked at the results and I can see the mismatch in units as well. Fixing this requires to change `ctuc = cu.prefactor_units(cc.ureg("kmol/m**3"), - dim)` into `ctuc = cu.prefactor_units(cc.ureg("kmol/m**3"), 1 - dim)` for the case with third body and no falloff.  I suspect that you applied the change only to `comp_qfqr` which is only used for some particular post processing routines. Instead, only `productionRate` is used when you do the chemistry integration. 

@jAnirudh  Can you please confirm that `productionRate` was not changed  in the test you ran?

By applying to change to `comp_qfqr` and `productionRate`, I actually see mismatch between Cantera and PelePhysics 

![1bar](https://user-images.githubusercontent.com/5544573/194179186-b3dbcda1-1af8-4ffa-9fc8-f2c598c84425.png)

This unit mismatch seems to be a Cantera vs PelePhysics convention. The third body concentration is computed in the same units as regular species. So we need to keep track of the units of the third body concentration which explains why our pre exponential factor compensates for the multiplication with regular reactant *and* third body concentration. Looking at the Cantera documentation, there seems to be something going on about incorporating the third body concentration to the reaction rate as well (see https://cantera.org/documentation/dev/sphinx/html/cython/utilities.html#cantera.use_legacy_rate_constants, under `cantera.use_legacy_rate_constants()`)

I would not consider this as a bug then. 

@jAnirudh  please let me know if I am missing something!

**Comment by jAnirudh** [comment]
@malihass thanks for looking into this. You're right this is not a bug.

If I understand correctly, PelePhysics is essentially following Cantera's legacy behaviour for the 3 body reaction rate constants. If I wanted to use the conventional formulation and define `ctuc = cu.prefactor_units(cc.ureg("kmol/m**3"), 1 - dim)` then I'd have to multiply `qdot` by the concentration of the third body, which, I'm assuming, is `1 m**3/kmol`.

To the first part about whether I modified both `comp_qfqr` and `productionRate`, I actually didn't change `ctuc` in either methods when I posted earlier. I was surprised that things were working despite `kf` being off by a factor of 1000, and was not able to figure out why that is the case. Once again, thanks for clearing things up for me :)
